### PR TITLE
AMQP-722 Support multiple routing keys in @QueueBinding

### PR DESCRIPTION
**Pull request by panchenko**
I described the idea in https://jira.spring.io/browse/AMQP-722

If we are going forward I could add javadocs to the affected privates methods and probably change Collection to Predicate, etc.

WDYT?

### CODE REVIEW

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessorTests.java [hunk 1]
```diff
@@ -195,6 +203,33 @@ public void invalidValueInAnnotationTestBean() {
...
+
+		RabbitListenerContainerTestFactory factory = context.getBean(RabbitListenerContainerTestFactory.class);
+		assertThat("one container should have been registered", factory.getListenerContainers(), hasSize(1));
+		RabbitListenerEndpoint endpoint = factory.getListenerContainers().get(0).getEndpoint();
+		assertEquals(Collections.singletonList("my_queue"),
+				((AbstractRabbitListenerEndpoint) endpoint).getQueueNames());
+		final List<Queue> queues = new ArrayList<>(context.getBeansOfType(Queue.class).values());
+		queues.sort(Comparator.comparing(Queue::getName));
+		assertThat(queues.stream().map(Queue::getName).collect(Collectors.toList()),
+				contains("my_queue", "secondQueue", "testQueue"));
```
Comment (panchenko) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessorTests.java]: This check can be simplified by moving `testQueue`/`secondQueue` from `Config` into another class, which is included only when needed. Can be a separate PR. Should I?

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessorTests.java [hunk 2]
```diff
@@ -269,6 +304,16 @@ public void handleIt(String body) {
 		}
 	}
 
+	@Component
+	static class MultipleBindingsTestBean {
+
+		@RabbitListener(bindings = {
+				@QueueBinding(exchange = @Exchange("my_exchange"), value = @org.springframework.amqp.rabbit.annotation.Queue(value = "my_queue", arguments = @Argument(name = "foo", value = "bar")), key = "red"),
+				@QueueBinding(exchange = @Exchange("my_exchange"), value = @org.springframework.amqp.rabbit.annotation.Queue("my_queue"), key = "yellow") })
```
Comment (panchenko) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessorTests.java]: Eclipse formatter insist on having it this way :) I can wrap it manually.

Comment (garyrussell) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessorTests.java]: Wouldn't it be a lot easier to change the key property to be an array...

	String[] key() default {};

... and then

		@RabbitListener(bindings = {
				@QueueBinding(exchange = @Exchange("my_exchange"), 
						value = @org.springframework.amqp.rabbit.annotation.Queue("my_queue"), 
							key = { "red", "yellow" }) })
		public void handleIt(String body) {
		}

?

**Comment by panchenko** [comment]
@garyrussell Your suggestion totally makes sense, I've updated this PR.

**Comment by garyrussell** [comment]
Merged as ddb42da8605bfcc287156da0b621383d391291db with minor polish and docs.

Thanks @panchenko !!

(Please limit commit comment headlines to 50 chars).
